### PR TITLE
Fix invalid CodeDom comment generation when a new line starts with a slash

### LINKED ISSUE #56267: CodeDom creates invalid C# line starting with `////` when a slash follows a newline in CodeComment.Text, causing warnings in Resources.Designer.cs files
Due to a bug in CodeDom's generation of C# code comments in both .NET Framework and .NET Core/5, any project with `<GenerateDocumentationFile>true</GenerateDocumentationFile>` that embeds a file that contains a line that starts with `/` gets invalid C# documentation syntax in .Designer.cs, resulting in warnings:

image

I asked first at https://github.com/dotnet/roslyn/issues/53729 in case the compiler team thought that documentation warnings should be skipped in generated code, and they decided that the current behavior is by design.

Then I filed an issue at https://github.com/dotnet/msbuild/issues/6677 since the StronglyTypedResourceBuilder which creates Resources.Designer.cs lives there. The same CodeDom problem afflicts the StronglyTypedResourceBuilder in .NET Framework's Microsoft.Build.Tasks.v4.0.dll assembly, file version 4.8.4084.0.

➡ The MSBuild team would like to know where this issue ranks for you.

### Details

This is where CodeDom creates the invalid `////` XML doc comment line: https://github.com/dotnet/runtime/blob/v5.0.6/src/libraries/System.CodeDom/src/Microsoft/CSharp/CSharpCodeGenerator.cs#L879
.NET Framework's CodeDom has the same bug: https://referencesource.microsoft.com/#System/compmod/microsoft/csharp/csharpcodeprovider.cs,e0b125d92a26ca23

On that line it outputs `///` because of `string commentLineStart = e.DocComment ? "///" : "//";` outside that loop. Then the next character of `value` that will be written after that is a single `/`. This causes invalid C# any time a forward slash follows a newline in `CodeComment.Text`.

### Steps to Reproduce

Full repro: [Repro.zip](https://github.com/dotnet/roslyn/files/6556702/Repro.zip)

Resource file being included (real-life example was a SQL file that was truncated by the resource code generator when included in the XML comment):

```

/
```

Problematic excerpt from Resources.Designer.cs:

```cs
/// <summary>
///   Looks up a localized string similar to 
////.
/// </summary>
internal static string SomeResourceFile {
    get {
        return ResourceManager.GetString("SomeResourceFile", resourceCulture);
    }
}
```

Project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
  </PropertyGroup>

  <ItemGroup>
    <Compile Update="Properties\Resources.Designer.cs">
      <DesignTime>True</DesignTime>
      <AutoGen>True</AutoGen>
      <DependentUpon>Resources.resx</DependentUpon>
    </Compile>
  </ItemGroup>

  <ItemGroup>
    <EmbeddedResource Update="Properties\Resources.resx">
      <Generator>ResXFileCodeGenerator</Generator>
      <LastGenOutput>Resources.Designer.cs</LastGenOutput>
    </EmbeddedResource>
  </ItemGroup>

</Project>
```

Specific errors:

image

image

image

### PR DESCRIPTION
**Pull request by jnm2**
Fixes https://github.com/dotnet/runtime/issues/56267

### CODE REVIEW

File: src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/CSharpCodeGenerationTests.cs [hunk 1]
```diff
@@ -3516,5 +3516,59 @@ public class ClassWithGenericMethod {
                   }
                 ");
         }
+
+        [Fact]
+        [ActiveIssue("https://github.com/dotnet/runtime/issues/56267", TargetFrameworkMonikers.NetFramework)]
```
Comment (jnm2) [inline src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/CSharpCodeGenerationTests.cs:3521]: This is what is done further up in this file on another test, using ActiveIssue to point to a closed GitHub issue.

Another option would potentially be `[SkipOnTargetFramework(TargetFrameworkMonikers.NetFramework, "https://github.com/dotnet/runtime/issues/56267")]`.

Comment (am11) [reply src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/CSharpCodeGenerationTests.cs:3521]: We have examples where there is reason description with link, such as:

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.ComponentModel.Annotations/tests/System/ComponentModel/DataAnnotations/MinLengthAttributeTests.cs#L75

**Comment by danmoseley** [comment]
Test failure is https://github.com/dotnet/runtime/issues/56649

**Comment by danmoseley** [comment]
LGTM, could you please change to SkipOnTargetFrameworks as suggested?

**Comment by jnm2** [comment]
Updated.

**Comment by jnm2** [comment]
@danmoseley I just realized there's no need to insert a space when generating an ordinary comment (e.g. `//`) and the line starts with more than one forward slash (resulting in `////`. Shall I update?

**Comment by danmoseley** [comment]
Up to you as long as there's tests 🙂

**Comment by jnm2** [comment]
Updated, ready for review again.

**Comment by jnm2** [comment]
@danmoseley I noticed that the filename wasn't updated to StringExtensions.cs, so I did that and squashed into my last commit which introduced the file.

**Comment by danmoseley** [comment]
Oops thanks

**Comment by jnm2** [comment]
GitHub showed that conflicts with main arose in the last week. I used a force push to fix this since no one has reviewed yet.

File: src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/VBCodeGenerationTests.cs [hunk 2]
```diff
@@ -3264,5 +3264,63 @@ End Sub
...
+        {
+            var codeTypeDeclaration = new CodeTypeDeclaration("ClassWithCommment")
+            {
+                IsClass = true,
+                Comments =
+                {
+                    new CodeCommentStatement(
+                        "' Lines starting with a single quote" + Environment.NewLine +
+                        "' each get a separating space," + Environment.NewLine +
+                        "but other lines do not get a space. This way generated files only change on tool upgrade where there were generation bugs.",
```
Comment (danmoseley) [inline src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/VBCodeGenerationTests.cs:3311]: I just noticed that this is invalid
```
''' <summary>
'''''starts with two quotes
''' </summary>
```
Ie a line starting with any number of quotes needs a space. I think your code does this, but we should include it in the test.

Comment (danmoseley) [reply src/libraries/System.CodeDom/tests/System/CodeDom/Compiler/VBCodeGenerationTests.cs:3311]: done

**Review by danmoseley — APPROVED** [review]

**Comment by danmoseley** [comment]
Unrelated failures. Thanks for the fix @jnm2 !

**Pull request merged**